packages/grafana-runtime/src/services/SidecarService_EXPERIMENTAL.ts

```diff
@@ -157,3 +166,97 @@ function getMainAppPluginId() {
 
   return mainApp || 'unknown';
 }
+
+type LocalStorageHistoryOptions = {
+  storageKey: string;
+};
+
+interface LocationStorageHistory extends H.MemoryHistory {
+  getLocationObservable(): Observable<H.Location | undefined>;
+}
+
+/**
+ * Simple wrapper over the memory history that persists the location in the localStorage.
+ *
+ * @param options
+ */
+function createLocationStorageHistory(options: LocalStorageHistoryOptions): LocationStorageHistory {
+  const storedLocation = localStorage.getItem(options.storageKey);
+  const initialEntry = storedLocation ? JSON.parse(storedLocation) : '/';
+  const locationSubject = new BehaviorSubject<H.Location | undefined>(initialEntry);
+  const memoryHistory = H.createMemoryHistory({ initialEntries: [initialEntry] });
+
+  let currentLocation = memoryHistory.location;
+
+  function maybeUpdateLocation() {
+    if (memoryHistory.location !== currentLocation) {
+      localStorage.setItem(
+        options.storageKey,
+        JSON.stringify(pick(memoryHistory.location, 'pathname', 'search', 'hash'))
+      );
+      currentLocation = memoryHistory.location;
+      locationSubject.next(memoryHistory.location);
+    }
+  }
+
+  // This creates a sort of proxy over the memory location just to add the localStorage persistence and the location
+  // observer. We could achieve the same effect by a listener but that would create a memory leak as there would be no
+  // reasonable way to unsubcribe the listener later on.
+  // Another issue is that react router for some reason does not care about proper `this` binding and just calls these
+  // as normal functions. So if this were to be a class we would still need to bind each of these methods to the
+  // instance so at that moment this just seems easier.
+  return {
+    get index() {
+      return memoryHistory.index;
+    },
+    get entries() {
+      return memoryHistory.entries;
+    },
+    canGo(n: number) {
+      return memoryHistory.canGo(n);
+    },
+    get length() {
+      return memoryHistory.length;
+    },
+    get action() {
+      return memoryHistory.action;
+    },
+    get location() {
+      return memoryHistory.location;
+    },
```

Comment:
Since we're not adding custom functionality to these methods, can we just do smth like `return {...memoryHistory, ...customisedMethods }` ? Some of these were removed in history v5 and we if we finally decide to upgrade, this will be an extra overhead.

Reply:
The thing is `getters` arent destructured as getter but as values at the moment of destructuring so I can get rid of some of the functions that way but most of this code still has to be there.

Reply:
What about overwriting only the methods we need, e.g., 
```ts
memoryHistory.push = (location: H.Path | H.LocationDescriptor<H.LocationState>, state?: H.LocationState) => {
  memoryHistory.push(location, state);
  maybeUpdateLocation();
};

return memoryHistory;
```

Reply:
But won't that issue be fixed by saving the original push into a variable? `const originalPush = memoryHistory.push.bind(memoryHistory);` 

Anyways, if refactoring this creates more issues than benefits, let's leave it as it is.